// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"team-manager/ent/predicate"
	"team-manager/ent/quote"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// QuoteDelete is the builder for deleting a Quote entity.
type QuoteDelete struct {
	config
	hooks    []Hook
	mutation *QuoteMutation
}

// Where appends a list predicates to the QuoteDelete builder.
func (qd *QuoteDelete) Where(ps ...predicate.Quote) *QuoteDelete {
	qd.mutation.Where(ps...)
	return qd
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (qd *QuoteDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, qd.sqlExec, qd.mutation, qd.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (qd *QuoteDelete) ExecX(ctx context.Context) int {
	n, err := qd.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (qd *QuoteDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(quote.Table, sqlgraph.NewFieldSpec(quote.FieldID, field.TypeInt64))
	if ps := qd.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, qd.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	qd.mutation.done = true
	return affected, err
}

// QuoteDeleteOne is the builder for deleting a single Quote entity.
type QuoteDeleteOne struct {
	qd *QuoteDelete
}

// Where appends a list predicates to the QuoteDelete builder.
func (qdo *QuoteDeleteOne) Where(ps ...predicate.Quote) *QuoteDeleteOne {
	qdo.qd.mutation.Where(ps...)
	return qdo
}

// Exec executes the deletion query.
func (qdo *QuoteDeleteOne) Exec(ctx context.Context) error {
	n, err := qdo.qd.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{quote.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (qdo *QuoteDeleteOne) ExecX(ctx context.Context) {
	if err := qdo.Exec(ctx); err != nil {
		panic(err)
	}
}
